### PR TITLE
Feature/game model

### PR DESCRIPTION
**Pull request by Kirschzelle**

### CODE REVIEW

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoard.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+class GameModelBoard : GameModelBoardInterface {
+
+    private var trump: String? = null
+    private var winningCard: GameModelCard? = null
+    private val game = ArrayList<GameModelCard>()
+    private val winningCards = ArrayList<GameModelCard>()
+    private val trumpStack = ArrayList<String>()
+
+    override fun addTrumpToStack(trump: String): GameModelResult<Unit> {
+        if ((trump.length != 2 || trump[0].code > 14 || (trump[1].code !in 1..4 && (trump[0].code != 0 || trump[1].code != 0)))) {
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoard.kt:12]: As personally discussed, this is a bit hard to follow in code. Let's create a future backlog task to work with more meaningful entities in code. For sending/receiving data it's fine to use very limited structure. But we don't need to rely on that structure at any other place.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoardInterface.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package at.aau.edu.wizards.gameModel
+
+interface GameModelBoardInterface {
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoardInterface.kt:3]: in kotlin interfaces are usually not prefixed by `Interface`. Usually this would be `GameModelBoard` and the implementation would be `GameModelBoardImpl`.

That also makes sense for repository pattern. Eg you define interface

```
interface UserRespostory {
   fun getUser(): User
}
```

implementation could look like:

`DatabaseUserRepository` or `InMemoryUserRepository` for specific cases. For default implementation it would be `UserRepositoryImpl`

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelDataListener.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        append(0.toChar())
+        append(0.toChar())
+    })
+    private var currentGuess = ArrayList<Guess>()
+    private var totalScores = ArrayList<Score>()
+
+    data class Guess(val guess: Int, val playerId: Int)
+    data class Score(val score: Int, val playerId: Int)
+
+    override fun getActivePlayer(): Int {
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/gameModel/GameModelDataListener.kt:19]: properties with public read and private set can be written

```
var property: Int = 0
    private set
``

**Comment by lechld** [comment]
All good, just minor comments.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoard.kt (hunk 1 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoard.kt]: Structure changed to a sealed class representation in new iteration

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoardInterface.kt (hunk 2 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelBoardInterface.kt]: Got rid of interfaces entirely in new implementation. Noted for future implementations.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelDataListener.kt (hunk 3 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelDataListener.kt]: Changed all private variables with public getters to public variables with private setters in new implementation.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                    return true
+                }
+                false
+            }
+            is GameModelResult.Success -> {
+                if (currentPlayer == id
+                    &&
+                    id == move[2].code
+                    && parent.listOfPlayers[id].cardsContain(move)
+                    && (move[0].code !in 1..13
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt]: Rewrote card into a sealed class of 4 different types, NoCard (Only used for internal purposes), Wizard(Color), Jester(Color), Normal(Color, Value) - Color can be of one of four types: Orange, Red, Blue, Green - In terms of future compatibility with UI representations I decided to leave Jester and Wizard cards with a color. (As to have something to differentiate them by.)

**Comment by Kirschzelle** [comment]
Code has been completely rewritten, review was fully integrated and performance increased. Security Hotspot is safe to use, but needs to be reviewed by someone with corresponding permissions. Capabilities were downgraded slightly, but it is now quite easy to implement any missing feature we might need later on.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        if (++dealer >= players.size) {
+            dealer = 0
+        }
+        currentPlayer = dealer
+        cardDealer.resetSet()
+    }
+
+
+    private fun endGame() {
+        //TODO
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt:55]: what about that? Shall we create tasks in backlog for such cases, idk?

**Review by lechld — APPROVED** [review]
Just check my comments. Feel free to merge then.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    fun playCard(card: GameModelCard) {
+        if (card is GameModelCard.NoCard) {
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt:59]: There are two reasons I decided to implement `NoCard`:

1: In wizards there is a special case where you try to deal over the card limit of 60 cards. Namely in the last round, when all cards are dealt between the players. If you then try to deal the card you would have no cards left, normally this would lead to an `ERROR`. Since the goal of this rework was to simplify the code I instead wrote the card dealer in a way that it returns a card of the instance `NoCard`, I can then go ahead and directly set this as the trump with the meaning: "There is no trump this round" instead of relying on more complex error handling to solve the issue.

2: I can avoid to use null-able variables in various parts of the code. Instead if there is no card for that variable at the moment you just set it to a card of instance `NoCard`. Avoiding any null casting, again improving readability.

File: app/src/main/java/at/aau/edu/wizards/sample/SampleViewModel.kt [hunk 7]
```diff
@@ -16,11 +16,11 @@ class SampleViewModel(
 
     init {
         // that launch allows us to call a function with "suspend"
-        viewModelScope.launch {
-            val receivedData = dataSource.provideData(someDependency)
+        //SampleViewModel.launch {
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/sample/SampleViewModel.kt:19]: These somehow must have sneaked in, i will revert this. Thanks for the note.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelPlayer.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    fun dealCards(amount: Int) {
+        for (card in 1..amount) {
+            cards.add(dealer.dealCardInSet())
+        }
+    }
+
+    fun hasColor(color: GameModelCard.Color): Boolean {
+        for (card in cards) {
+            when (card) {
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelPlayer.kt:23]: Personally, I don't think loc (lines of code) is too useful of a grading tool. For example this code was completely revised 3 times and that doesn't show up as jack shit in loc. I think the point here is just to make sure everyone actually does a considerable part of the finished project.

I think the one-liner is considerably more readable. It is written this way because I didn't know you could write it that way. I will go trough the code again and try to switch out similar cases.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt (hunk 5 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt:55]: Well, if we refer to our prototype the score screen would come after, so we should switch to the `Fragment` of the score screen. However since we do this inside the fragment we should communicate to this, which in this case I have not done. So actually this is a bit of an oversight from me.

I will implement a call to `sendMove` of the `GameModel`, that sends some String like: "GameEnded", that we can then catch later on in the Fragment without the need to go back into the Logic.

File: app/src/test/java/at/aau/edu/wizards/gameModel/GameModelRulesUnitTest.kt [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        assert(rules.localPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Red, 9)))
+        assertFalse(rules.localPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Orange, 10)))
+        assert(rules.localPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Blue, 7)))
+        assertFalse(rules.localPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Red, 7)))
+        assertFalse(rules.localPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Orange, 11)))
+
+        assertFalse(rules.currentPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Red, 9)))
+        assert(rules.currentPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Orange, 10)))
+        assertFalse(rules.currentPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Blue, 7)))
+        assert(rules.currentPlayerOwns(GameModelCard.Normal(GameModelCard.Color.Red, 7)))
```
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/test/java/at/aau/edu/wizards/gameModel/GameModelRulesUnitTest.kt:85]: Noted, using different libraries is not ideal. We could decide to do an !optional! backtracking task to keep track of it, but it is not really worth it to rewrite from a time perspective. If someone else feels like fixing this, feel free to do so.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt (hunk 5 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelRules.kt:55]: There is also a second TODO, but that is marking the interface between networking and the model and as such needs to be implemented with the caller in mind (game fragment). I do not think it necessary to create a backlog task for this.

No further cases exist.

File: app/src/main/java/at/aau/edu/wizards/gameModel/GameModelPlayer.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/gameModel/GameModelPlayer.kt:23]: Including above line, found 2 similar cases and changed them.
